### DISCUSSION TITLE
Is there anyway to have `KeepPreviousData` as function of the queryKey instead of being a boolean

### DISCUSSION THREAD

**Discussion by tayelno** [TanStack/query#4828] [Q&A] [answered]
Can we have something like the following or is it already achievable through other configurations?

The idea here is that I have multiple properties in my queryKey but if `propertyThatIfChangedIWantToShowLoadingState` change I want the query not to show the previous data, as for this specific property I want my flow to be:

loaded -> loading -> loading or error
                               
instead of

loaded -> loaded (which is what keepPreviousData typically translates to)

```typescript

useQuery({
    queryFn,
    queryKey,
    keepPreviousData: (key: QueryKey) => {
      if (key.propertyThatIfChangedIWantToShowLoadingState) {
          return false
      }
      return true
    }
  });

```

**Comment by TkDodo** [comment, marked as answer]
Does this not work? You should be able to toggle keepPreviousData on and off as you please:

```
useQuery({
    queryFn,
    queryKey,
    keepPreviousData: queryKey.propertyThatIfChangedIWantToShowLoadingState,
  });
```

---

Please note that for v5, we have already removed `keepPreviouData` and merged it with `placeholderData`:

```
useQuery({
  queryKey,
  queryFn,
  placeholderData: (previousData) => {}
})
```

you'll get the previous data passed in and can decide if you want to "keep" showing it as a placeholder for the current query key.

what we could potentially do is also send the whole `query` in as a second parameter, but just as before, you also have access to the key in your closure, so I don't think it would be necessary.

You can read more on that in the RFC here: https://github.com/TanStack/query/discussions/4426

**Reply by tayelno** [reply]
my solution was implementing a usePreviousValue hook that keeps the last instance of that queryKey param `meta.duration` in example

```typescript

export function useKPIWidgetData(meta: WidgetMetadata, networkId: string) {
  const queryKey = [meta.metric, meta.duration, networkId];
  const previousDuration = usePreviousValue(meta.duration); // save previous key value
  async function queryFn() {
    const data= await fetchData({ meta, networkId });
    return data;
  }
  const query = useQuery({
    queryFn,
    queryKey,
    refetchInterval: THREE_MINUTES,
    staleTime: THREE_MINUTES,
    keepPreviousData: previousDuration !== meta.duration, // use it for comparison here
  });

  return query;
}

```

Using placeholdeData is a better solution, but can `placeholderData` pass the previousQuery as well ?

**Reply by TkDodo** [reply]
it could I think, but what would you need it for?

**Reply by tayelno** [reply]
in my case it would be something like

```typescript
export function useKPIWidgetData(meta: WidgetMetadata, networkId: string) {
  const queryKey = { duration: meta.duration, size: meta.size }


  async function queryFn() {
    const data= await fetchData({ meta, networkId });
    return data;
  }



  function placeholderData(previousData, previousQuery) {
    // I want the data to be refetched whenever one previous queryKey props get changed
   // but only keep previous data if `duration` prop did not change
    if (previousQuery.duration === queryKey.duration) {
      return previousData
    }
    return undefined
  }



  const query = useQuery({
    queryFn,
    queryKey,

   placeholderData
  });

  return query;
}
```

**Reply by tayelno** [reply]
Not sure how would we go back to a `loading` state using `placeholderData` ?

**Reply by TkDodo** [reply]
couldn't you use the queryKey from the closure - just accessing `meta.duration` instead of `queryKey.duration` ?

**Reply by tayelno** [reply]
problem is that I need the `previousQuery.duration` not the current value
